HIVE-21867

jcamachor opened the pull request.

vineetgarg02 commented on ql/src/test/results/clientpositive/llap/hybridgrace_hashjoin_2.q.out:
```diff
@@ -1421,7 +1421,7 @@ STAGE PLANS:
                           input vertices:
                             1 Map 5
-                          Statistics: Num rows: 25 Data size: 2225 Basic stats: COMPLETE Column stats: COMPLETE
+                          Statistics: Num rows: 4 Data size: 356 Basic stats: COMPLETE Column stats: COMPLETE
```
I wonder why stats estimation changed

jcamachor replied:
Good catch, thanks! This is unexpected indeed... Taking a look now.

vineetgarg02 commented on ql/src/java/org/apache/hadoop/hive/ql/parse/TezCompiler.java:
```diff
@@ -1766,6 +1774,59 @@ private void removeSemijoinOptimizationByBenefit(OptimizeTezProcContext procCtx)
       GenTezUtils.removeSemiJoinOperator(procCtx.parseContext, rs, ts);
     }
+
+    for (Entry<FilterOperator, Collection<SemijoinOperatorInfo>> e : globalReductionFactorMap.asMap().entrySet()) {
```
Creating separate method for this and adding comments to explain why we are doing it will be nice.

jcamachor replied:
Good idea, will do.

vineetgarg02 commented on ql/src/test/results/clientpositive/llap/mergejoin.q.out:
```diff
@@ -41,8 +41,8 @@ STAGE PLANS:
                         className: VectorFilterOperator
                         native: true
-                        predicateExpression: FilterExprAndExpr(children: SelectColumnIsNotNull(col 0:string), FilterExprAndExpr(children: FilterStringColumnBetweenDynamicValue(col 0:string, left NULL, right NULL), VectorInBloomFilterColDynamicValue))
-                    predicate: (key is not null and (key BETWEEN DynamicValue(RS_7_b_key_min) AND DynamicValue(RS_7_b_key_max) and in_bloom_filter(key, DynamicValue(RS_7_b_key_bloom_filter)))) (type: boolean)
```
Positive change :thumbsup:

jcamachor commented:
@vineetgarg02 , I updated the PR. Note that ```hybridgrace_hashjoin_2.q``` issue will be tackled in HIVE-20260 so I think we can proceed with this issue. In the latest patch I also added some simple logic to ```SharedWorkOptimizer``` to remove some duplicate filter expressions remaining in the plan because of change in the shape of the expression (problem was existing, patch just exposed it).
